rust/agents/relayer/src/relayer.rs

```diff
@@ -98,52 +98,72 @@ impl Relayer {
         sync
     }
 
-    fn run_inbox(&self, inbox_contracts: InboxContracts) -> Instrumented<JoinHandle<Result<()>>> {
-        let db = self.outbox().db();
-        let checkpoint_relayer = CheckpointRelayer::new(
+    fn run_checkpoint_fetcher(
+        &self,
+        signed_checkpoint_sender: Sender<Option<MultisigSignedCheckpoint>>,
+    ) -> Instrumented<JoinHandle<Result<()>>> {
+        let checkpoint_fetcher = CheckpointFetcher::new(
             self.outbox().outbox(),
-            self.polling_interval,
-            self.submission_latency,
-            self.relayer_message_processing,
-            db.clone(),
-            inbox_contracts.clone(),
+            self.signed_checkpoint_polling_interval,
             self.multisig_checkpoint_syncer.clone(),
+            signed_checkpoint_sender,
             self.core.metrics.last_known_message_leaf_index(),
         );
+        checkpoint_fetcher.spawn()
+    }
+
+    fn run_inbox(
+        &self,
+        inbox_contracts: InboxContracts,
+        signed_checkpoint_receiver: Receiver<Option<MultisigSignedCheckpoint>>,
+    ) -> Instrumented<JoinHandle<Result<()>>> {
+        let db = self.outbox().db();
         let message_processor = MessageProcessor::new(
             self.outbox().outbox(),
-            self.polling_interval,
-            self.max_retries,
+            self.max_processing_retries,
             db,
-            self.submission_latency,
-            inbox_contracts.inbox,
+            inbox_contracts,
+            signed_checkpoint_receiver,
             self.core.metrics.last_known_message_leaf_index(),
             self.core.metrics.retry_queue_length(),
         );
 
-        self.run_all(vec![checkpoint_relayer.spawn(), message_processor.spawn()])
+        message_processor.spawn()
     }
 
     fn wrap_inbox_run(
```

Comment:
nit: this is Rust not C++ so we should put more private functions that are depended on by public ones after generally. Probably not worth changing for this PR.